Local date range

PeterAttardo opened the pull request:
This pull request enables Kotlin range and progression semantics with the LocalDate and DatePeriod classes. Its implementation borrows from the patterns used for `IntRange`.

Example:

```
val start = LocalDate.parse("2020-01-01")
val end = LocalDate.parse("2020-12-31")

// calls function for each day of the year in 2020
(start..end).forEach {
    processDayOf2020(it)
}

// calls function for each day of the year in 2020, ordered in reverse
(end downTo start).forEach {
   processDayOf2020(it)
}

// calls function for every other day of the year in 2020
(start..end step DatePeriod(days=2)).forEach {
   processDayOf2020(it)
}
```

dkhalanskyjb commented:
We have a similar request, but for `Instant`: https://github.com/Kotlin/kotlinx-datetime/issues/34
Could you please describe why this is needed? Either in the issue I linked if your use cases are similar to the ones listed there, or in a new issue.

boswelja commented:
@dkhalanskyjb I've had a use case for this, though it's more of a convenience having a dedicated `LocalDateRange` compared to a `ClosedRange<LocalDate>`. I'm building a calendar library that allows consumers to update a particular range of dates on the screen, the intended use case for this is to allow changing the rendered content in chunks.

A few other use cases off the top of my head:

* Representing a selected date range
* Using plus/minus + forEach in conjunction to iterate over a large range in chunks (granted this could be achieved by having a list of dates and chunking that, but if you've only got a start and end date this could be more efficient)

dkhalanskyjb commented on core/common/src/LocalDateRange.kt:
```diff
@@ -0,0 +1,118 @@
+        start: LocalDate,
+        endInclusive: LocalDate,
+        step: Int
+    ) : this(IntProgression.fromClosedRange(start.toEpochDays(), endInclusive.toEpochDays(), step))
```
This implementation is incorrect for dates whose `toEpochDays()` does not fit into an `Int`. See https://kotlinlang.org/api/kotlinx-datetime/kotlinx-datetime/kotlinx.datetime/-local-date/to-epoch-days.html

dkhalanskyjb replied:
Why? Using `IntProgression` looks like just an implementation detail, not some fundamental issue. It would be better to just support all dates without relying on `toEpochDays()`.

PeterAttardo replied:
It looks like the bounds of `LocalDate` vary by implementation (`native` goes from year `-999_999` to `999_999`, while `jvm` goes from year `-999_999_999` to `999_999_999`) but each internally convert to epoch days in order to do arithmetic. However those internal conversions to epoch days can be broader than the `common` conversion to epoch days. I do wonder if it would be better for the `common` `toEpochDays()` to return a `Long` instead of an `Int`, so that it is not more restricted than some implementations of `LocalDate` itself. 

An alternative implementation is to switch to a `LongProgression` between `0` and `size of progression` and use `fun next() = startDate.plus(iterator.next(), DateTimeUnit.DAY)`, which would bypass the `common` `toEpochDays()` and `fromEpochDays()`. However `daysUntil` _also_ returns an `Int`, and so the size of the range can't be reliably computed either. Without the ability to compute the size of the progression, we also lose the ability to generate random values from within the progression without doing something hideous like materializing the entire progression. 

I could attempt to reimplement it without leaning on any of the primitive progressions at all, but we would still lose the ability to calculate the last element (again excepting a hideous full iteration) for steps other than 1 day and -1 day, as we can't calculate the `differenceModulo` without an `Int`/`Long` representation of `LocalDate`. 

It seems like our options are:

1. Change the return type of several `common` methods from `Int` to `Long` to ensure that the full range of `LocalDate` can be expressed as days without overflowing.
2. Accept that the use of `Int` in `common` clamps the allowable ranges tighter than some implementations of `LocalDate` itself.
3. Push everything down and implement `LocalDateProgression` on a per-platform basis.

I'm inclined to leave it clamped for now, and then unclamp it in the future if/when the `common` day methods are upgraded from `Int` to `Long`.

dkhalanskyjb replied:
> I do wonder if it would be better for the common toEpochDays() to return a Long instead of an Int, so that it is not more restricted than some implementations of LocalDate itself.

Maybe it would, but that's a separate problem. We're considering making the ranges the same, and it's possible that `toEpochDays()` will naturally have to return a `Long`: https://github.com/Kotlin/kotlinx-datetime/issues/432

> However `daysUntil` also returns an Int, and so the size of the range can't be reliably computed either.

Inside the library, we're not bound by the `public` API. If you need a `daysUntil` that returns a `Long`, you can implement it using something like

```
// common code
internal expect fun LocalDate.daysUntilLong(other: LocalDate): Long

// JVM
internal actual fun LocalDate.daysUntilLong(other: LocalDate): Long =
    this.value.until(other.value, ChronoUnit.DAYS)

// other platforms
internal actual fun LocalDate.daysUntilLong(other: LocalDate): Long =
    daysUntil(other).toLong()
```

PeterAttardo replied:
I added platform-specific `toEpochDaysLong()` and `fromEpochDays(epochDays: Long)` functions, and switched from an `IntProgression` to a `LongProgression`. So far, so good. The one issue that remains is that, when implementing `Collection<LocalDate>`, it expects `size` to be an `Int`. Because we have expanded `LocalDateProgression` beyond the bounds of `Int`, it is now possible to create progressions of greater size than `Int.MAX_VALUE`.

In light of this, do we still want to implement `Collection`, and if so, do we want to throw in these cases, or clamp the size to `Int`?

dkhalanskyjb replied:
I think we should clamp the size to `Int.MAX_VALUE` for consistency with Java (and, most likely, the other implementations of `Collection` we have with the same problem): https://docs.oracle.com/en/java/javase/17/docs/api/java.base/java/util/Collection.html#size() @ilya-g, WDYT?

dkhalanskyjb replied:
It turns out we (the Kotlin library team) have already decided to use the same contract as Java here, I just didn't know that.

dkhalanskyjb commented on core/common/src/LocalDateRange.kt:
```diff
@@ -0,0 +1,118 @@
+import kotlin.random.Random
+import kotlin.random.nextInt
+
+internal class LocalDateProgressionIterator(private val iterator: IntIterator) : Iterator<LocalDate> {
```
I think this can be a `private class`.

dkhalanskyjb commented on core/common/src/LocalDateRange.kt:
```diff
@@ -0,0 +1,118 @@
+
+
+    public companion object {
+        public fun fromClosedRange(
```
Until it's clear that this function is useful, it can be kept `internal`.

dkhalanskyjb replied:
IIRC, we don't expect this function to be useful to the public (and it looks like we're right not to: https://grep.app/search?q=.fromClosedRange%28), and instead, it's there for the compiler to generate more optimized code when people write things using the functions we do expect them to use. So, `((1..11) step 2).reversed()` becomes something like `fromClosedRange(11, 1, -2)`. See https://github.com/JetBrains/kotlin/blob/658b9e9a4b11171bd09127113e152c12aaa5597d/libraries/tools/kotlin-stdlib-gen/src/templates/Ranges.kt `LocalDate`, being outside the standard library, can't benefit from this.

@ilya-g knows more about this and could maybe correct me.

PeterAttardo replied:
I'll switch it to internal

dkhalanskyjb commented on core/common/src/LocalDateRange.kt:
```diff
@@ -0,0 +1,118 @@
+
+    override fun equals(other: Any?): Boolean {
+        if (this === other) return true
+        if (other == null || this::class != other::class) return false
```
```suggestion
        if (other !is LocalDateProgression) return false
```
Then you won't need the following line, and in fact, this whole function can be a perfectly readable one-liner.

dkhalanskyjb commented on core/common/src/LocalDateRange.kt:
```diff
@@ -0,0 +1,118 @@
+            rangeEnd: LocalDate,
+            stepValue: Int,
+            stepUnit: DateTimeUnit.DayBased
+        ): LocalDateProgression = LocalDateProgression(rangeStart, rangeEnd, stepValue * stepUnit.days)
```
Multiplication can overflow an `Int`. This must be accounted for.

dkhalanskyjb commented on core/common/src/LocalDateRange.kt:
```diff
@@ -0,0 +1,118 @@
+public class LocalDateRange(start: LocalDate, endInclusive: LocalDate) : LocalDateProgression(start, endInclusive, 1), ClosedRange<LocalDate>, OpenEndRange<LocalDate> {
+    override val start: LocalDate get() = first
+    override val endInclusive: LocalDate get() = last
+    override val endExclusive: LocalDate get() = endInclusive.plus(1, DateTimeUnit.DAY)
```
See the proposed API: this function should be deprecated, as it can throw an exception on overflow, which may be surprising. Also, this function should include the documentation and throw `IllegalStateException` if the overflow did happen. See https://kotlinlang.org/api/latest/jvm/stdlib/kotlin.ranges/-char-range/end-exclusive.html

PeterAttardo replied:
Will add check and deprecation warning

dkhalanskyjb commented on core/common/src/LocalDateRange.kt:
```diff
@@ -0,0 +1,118 @@
+
+public fun LocalDateProgression.reversed(): LocalDateProgression = LocalDateProgression(intProgression.reversed())
+
+public fun LocalDateProgression.step(value: Int, unit: DateTimeUnit.DayBased) : LocalDateProgression = LocalDateProgression(intProgression.step(value * unit.days))
```
Arithmetic overflow.

dkhalanskyjb commented on core/common/src/LocalDateRange.kt:
```diff
@@ -0,0 +1,118 @@
+public operator fun LocalDate.rangeTo(that: LocalDate): LocalDateRange = LocalDateRange(this, that)
+public operator fun LocalDate.rangeUntil(that: LocalDate) : LocalDateRange = rangeTo(that.minus(1, DateTimeUnit.DAY))
+
+public fun LocalDateProgression.random(randomIntFunction: (IntRange) -> Int) : LocalDate = intProgression.random(randomIntFunction)
```
Are you certain we need these functions? I can't think of a use case for them.

PeterAttardo replied:
Most use cases will be fine with selecting from a uniform distribution, but any time random selection is at play, I assume there are some use cases for selecting from a normal distribution, a Poisson distribution, etc. This function allows a user to implement whatever distribution logic they want.

PeterAttardo replied:
I'll remove the function and let advanced users lean on a custom implementation of `Random`. 

Outside the scope of this library, but this does make me interested in formal implementations of `NormalRandom`, `PoissonRandom`, `BinomialRandom`, etc.

dkhalanskyjb commented on core/common/test/DateTimePeriodTest.kt:
```diff
@@ -149,9 +149,19 @@ class DateTimePeriodTest {
         assertTrue(dp2 is DatePeriod)
+        assertTrue(dp3 is DatePeriod)
+        assertEquals(DateTimePeriod(years = -10, days=-3, hours = -2), -(p1 + p2 + p3))
+        assertEquals(DatePeriod(years = -11, months = -6), -(dp1 + p1))
```
It doesn't seem like these tests are still relevant to the PR.

PeterAttardo replied:
They're from the changes to `DatePeriod` arithmetic, which I'll be reverting. I'll revert these as well.

dkhalanskyjb commented on core/common/src/LocalDateRange.kt:
```diff
@@ -0,0 +1,118 @@
+
+    public open fun isEmpty(): Boolean = intProgression.isEmpty()
+
+    override fun toString(): String = if (intProgression.step > 0) "$first..$last step ${intProgression.step}" else "$first downTo $last step ${intProgression.step}"
```
If you call `toString` on other range implementations, the result can be copy-pasted into the code to obtain the same progression. This one can not, and I think it would be nice to preserve that property.

PeterAttardo replied:
This is the one place where `step` is semi-exposed publicly. Since step isn't really settled, I didn't want to commit too much in any direction.

Being able to copy-paste into code is a nice consequence of the code representation of primitives matching their string representation, but they diverge for more complex objects. Additionally, there is no infix step function for `LocalDateProgression` yet. I think we would lose a lot in readability if we were to make it copy-pasteable:

`LocalDate.parse("2000-01-01")..LocalDate.parse("2000-01-10").step(2, DateTimeUnit.DAY)` is a pretty ugly string representation. We could improve it slightly with:

`LocalDate(2000, 1, 1)..LocalDate(2000, 1, 10).step(2, DateTimeUnit.DAY)`, but it still has poor readability, and at that point we're abandoning the existing string representation of `LocalDate` entirely.

dkhalanskyjb replied:
Ok, makes sense, thanks! Could you then include the information that the step is measured in days?

dkhalanskyjb commented on core/common/src/LocalDateRange.kt:
```diff
@@ -0,0 +1,118 @@
+public infix fun LocalDate.downUntil(that: LocalDate) : LocalDateProgression = downTo(that.plus(1, DateTimeUnit.DAY))
+
+public operator fun LocalDate.rangeTo(that: LocalDate): LocalDateRange = LocalDateRange(this, that)
+public operator fun LocalDate.rangeUntil(that: LocalDate) : LocalDateRange = rangeTo(that.minus(1, DateTimeUnit.DAY))
```
This can overflow and throw, and this can be avoided.

PeterAttardo replied:
Looking to understand how best to avoid the overflow. If I'm understanding correctly, this can only overflow if `that` is `LocalDate.MIN` (every other date has a valid `LocalDate` one day before it). By definition this would make the range empty, as every possible `this` that yields a non-empty range would itself be less than `LocalDate.MIN` and therefore not valid. This would seem to imply you can just return `LocalDateRange.EMPTY` any time `that` is `LocalDate.MIN`, however `EMPTY` would have incorrect values for `start` and `endInclusive`, which are part of the public API. Even if you preserved `start` by returning `this.rangeTo(LocalDate.MIN)`, your `endInclusive` would still be one day off, and if `this` happened to be `LocalDate.MIN` you'd accidentally create a non-empty range. I'm not sure how to avoid the throw while still maintaining correctness in the resulting range.

dkhalanskyjb replied:
As a rule of thumb, when in doubt, just try to keep the behavior consistent with the stdlib:
```kotlin
fun main() {
    val range = 10 until (Int.MIN_VALUE)
    println(range) // prints "1..0"
}
```
So it's okay to return `EMPTY` if you notice you can't do anything meaningful and the range *is* empty. When it's possible to preserve the boundaries, then sure, it's worth doing:
```kotlin
fun main() {
    val range = 10 until (-1)
    println(range) // prints "10..-2"
}
```

dkhalanskyjb commented on core/common/src/LocalDateRange.kt:
```diff
@@ -0,0 +1,118 @@
+public fun LocalDateProgression.step(value: Long, unit: DateTimeUnit.DayBased) : LocalDateProgression = step(value.toInt(), unit)
+
+public infix fun LocalDate.downTo(that: LocalDate) : LocalDateProgression = LocalDateProgression.fromClosedRange(this, that, -1, DateTimeUnit.DAY)
+public infix fun LocalDate.downUntil(that: LocalDate) : LocalDateProgression = downTo(that.plus(1, DateTimeUnit.DAY))
```
I don't think we have a `downUntil` function for any other ranges. Consistency of experience is extremely important to us: we want our users' knowledge to be completely transferable across the API entries we provide. "Aha, I need an `Int` range, and I already know how to do this for `LocalDate` ranges" should always work, unless there's a significant barrier to that.

Also, overflow.

PeterAttardo replied:
I added it for symmetry with `rangeUntil`, but I can remove it.

dkhalanskyjb reviewed (changes requested):
* Could you also make the range an implementation of `Collection`? See https://youtrack.jetbrains.com/issue/KT-36932/Let-Progression-classes-implement-Collection-to-reduce-time-when-iterates-over-IntRange-when-cast-to-Iterable
* All API entries (classes, functions, properties) need separate tests, a KDoc, and a usage sample.

PeterAttardo commented:
Something I came across that's not exactly an issue in the library, but is a usability question when using the library in IntelliJ that I'm wondering if there's a good answer for.

Because Kotlin automatically imports `kotlin.ranges.*` into every file, that means that the extension function `fun <T : Comparable<T>> T.rangeTo(that: T): ClosedRange<T>` is by-default available for all code. `LocalDate` implements `Comparable<LocalDate>` and so if you type the following:

```Kotlin
(localDate1..localDate2).forEach {
    // some code in here
}
```

`localDate1..localDate2` will **not** display an `Unresolved reference` warning due to a failure to import `kotlinx.datetime.rangeTo`, but will happily use `kotlin.ranges.rangeTo` and return a `ClosedRange<LocalDate>` instead. Because `ClosedRange<LocalDate>` is not `LocalDateRange`, all of our nice progression methods are unavailable, and we get an `Unresolved reference` warning on `forEach` instead.

This is easily solved by importing `kotlinx.datetime.rangeTo`, but IntelliJ is unable to recognize this as the solution, because from its perspective there is no issue resolving the `rangeTo` operator, just an issue with calling non-existing methods on `ClosedRange<LocalDate>`. This means users will have to be aware of this pitfall, and manually type in the import in every file they want to use the range operator with `LocalDate`.

dkhalanskyjb commented:
@PeterAttardo, maybe this issue can be solved if `rangeTo` is a member method of `LocalDate`? Good job noticing this problem!

PeterAttardo commented:
@dkhalanskyjb I was hoping not to have to touch the `LocalDate` class, but that does seem like the cleanest solution. I have added the range operators to `LocalDate`, but because expect classes don't allow default implementations, I have moved the implementation logic into static methods in `LocalDateRange`. That can be cleaned up in the future if default implementations are ever allowed.

dkhalanskyjb commented:
`@ilya-g` observed that the issue of forgetting to import methods actually affects the other extension functions in this PR as well: `last()` on a `LocalDateRange` shouldn't be the one used for generic collections. Still, making all these extension methods into members just for that wouldn't be correct: they *are* extensions semantically, and there are a lot of them. In the end, we decided this was a tooling problem: https://youtrack.jetbrains.com/issue/KTIJ-31281

Still, adding the range-creating methods directly to `LocalDate` makes sense semantically, so I think we should stick to that.

PeterAttardo commented:
@dkhalanskyjb Sounds like a good solution to address this with tooling. I focused on the range operators because, while you're right that imports can be overlooked for any more-specific extension method, at least with named methods you can type the first few letters, and then select the appropriate extension from the drop down, which will then auto-import the extension method. There's no such tooling for operators, so it's a manual type or copy-paste every time (or I suppose using the `rangeTo` and `rangeUntil` nomenclature to force the import, and then rewrite to the operator notation, which is equally arduous).

PeterAttardo commented on core/common/test/LocalDateRangeTest.kt:
```diff
@@ -0,0 +1,274 @@
+        }
+        repeat(20) {
+            assertTrue {
+                (Jan_01_2000..Jan_24_2000).step(5, DateTimeUnit.DAY).let { it.contains(it.random()) }
```
This test is not deterministic, which I'm not thrilled about. The other way to test it would be with a spy, but adding mockk just for this one test case seemed like overkill/overstepping.

dkhalanskyjb replied:
In my opinion, this test is just fine the way it is, but if you prefer to test this deterministically, you can make a custom `Random` implementation and pass it to the function.

dkhalanskyjb replied:
In my opinion, this test is fine the way it is (in fact, we have other randomly generated things in our tests), but if you prefer to test this deterministically, you can make a custom `Random` implementation and pass it to the function.

dkhalanskyjb commented:
@PeterAttardo, most discussions are marked as resolved; looks like most is this ready for a new round of reviews? If so, please click "re-request review" in the "Reviewers" panel.

dkhalanskyjb commented on core/commonJs/src/LocalDate.kt:
```diff
@@ -42,6 +42,13 @@ public actual class LocalDate internal constructor(internal val value: jtLocalDa
             throw e
         }
 
+        internal actual fun fromEpochDays(epochDays: Long): LocalDate {
+            require(epochDays in Int.MIN_VALUE..Int.MAX_VALUE) {
+                "Invalid date: boundaries of LocalDate exceeded"
+            }
+            return fromEpochDays(epochDays.toInt())
+        }
```
```suggestion
        internal actual fun fromEpochDays(epochDays: Long): LocalDate =
            fromEpochDays(epochDays.clampToInt())
```
Our internal `clampToInt` function clamps the `Long` to `Int.MAX_VALUE` or `Int.MIN_VALUE`, and both of these values don't fit into the range checked by `fromEpochDays(Int)`.

PeterAttardo replied:
> both of these values don't fit into the range checked by fromEpochDays(Int)

We can assume this to be true because we know how `fromEpochDays(Int)` is implemented, but the signature itself does not require that to be the case. Do we mind that we're relying on that assumption?

dkhalanskyjb replied:
`fromEpochDays(Long)` already has the knowledge that days outside `Int` won't work on this platform, and learning the same about `Int.MAX_VALUE` is just the next step.

dkhalanskyjb commented on core/commonKotlin/src/LocalDate.kt:
```diff
@@ -82,6 +82,13 @@ public actual class LocalDate actual constructor(public actual val year: Int, pu
             return LocalDate(yearEst, month, dom)
         }
 
+        public actual fun fromEpochDays(epochDays: Long): LocalDate {
+            require(epochDays in MIN_EPOCH_DAY..MAX_EPOCH_DAY) {
+                "Invalid date: boundaries of LocalDate exceeded"
+            }
+            return fromEpochDays(epochDays.toInt())
+        }
```
```suggestion
        internal actual fun fromEpochDays(epochDays: Long): LocalDate =
            fromEpochDays(epochDays.clampToInt())
```
First, `clampToInt`, like above. Second: `public` seems incorrect.

dkhalanskyjb commented on core/commonKotlin/src/LocalDate.kt:
```diff
@@ -121,6 +128,10 @@ public actual class LocalDate actual constructor(public actual val year: Int, pu
         return total - DAYS_0000_TO_1970
     }
 
+    internal actual fun toEpochDaysLong() : Long {
+        return toEpochDays().toLong()
+    }
```
```suggestion
    internal actual fun toEpochDaysLong(): Long = toEpochDays().toLong()
```
Style: consistency with the other file, no space before `:`.

dkhalanskyjb commented on core/jvm/src/LocalDate.kt:
```diff
@@ -39,6 +39,9 @@ public actual class LocalDate internal constructor(internal val value: jtLocalDa
         public actual fun fromEpochDays(epochDays: Int): LocalDate =
             LocalDate(jtLocalDate.ofEpochDay(epochDays.toLong()))
```
Minor: I'd write this as `fromEpochDays(epochDays.toLong())` to avoid the tiny bit of unnecessary code duplication. Maybe this goes for `toEpochDays()`, though there, the benefit is even more miniscule. Up to you.

dkhalanskyjb commented on core/common/test/samples/LocalDateRangeSamples.kt:
```diff
@@ -0,0 +1,58 @@
+        // Creating LocalDateRange from LocalDate
+        check(Jan_01_2000..Jan_05_2000 == LocalDateRange(Jan_01_2000, Jan_05_2000))
+        check(Jan_01_2000..<Jan_06_2000 == LocalDateRange(Jan_01_2000, Jan_05_2000))
+        check(Jan_05_2000 downTo Jan_01_2000 == LocalDateProgression(Jan_05_2000, Jan_01_2000, -1))
```
Samples shouldn't contain internal API (like `LocalDateProgression` constructor here). Suggestion: how about something like `check((LocalDate(1, 1, 2000)..LocalDate(1, 5, 2000)).toList() == listOf(...))` instead? It's verbose but explains in no uncertain terms what ranges contain, which is what we want from samples.

dkhalanskyjb commented on core/common/test/samples/LocalDateRangeSamples.kt:
```diff
@@ -0,0 +1,58 @@
+class LocalDateRangeSamples {
+    val Jan_01_2000 = LocalDate(2000, 1, 1)
+    val Jan_05_2000 = LocalDate(2000, 1, 5)
+    val Jan_06_2000 = LocalDate(2000, 1, 6)
```
The samples are broken, as they will not include these: 
![image](https://github.com/user-attachments/assets/93b2d1fe-0aa3-4129-a72f-ba7fa564cc27)
How to check this: `./gradlew dokkaHtml`, then open `./core/build/dokka/html/kotlinx-datetime/kotlinx.datetime/-local-date/range-to.html`

dkhalanskyjb commented on core/common/src/LocalDateRange.kt:
```diff
@@ -0,0 +1,263 @@
+ */
+public infix fun LocalDate.downTo(that: LocalDate) : LocalDateProgression = LocalDateProgression.fromClosedRange(this, that, -1, DateTimeUnit.DAY)
+
+/**
+ * Returns a random [LocalDate] within the bounds of the [LocalDateProgression].
+ *
+ * Takes the step into account; will not return any value within the range that would be skipped over by the progression.
+ *
+ * @throws IllegalArgumentException if the progression is empty.
+ *
+ * @sample kotlinx.datetime.test.samples.LocalDateRangeSamples.random
+ */
+public fun LocalDateProgression.random(random: Random = Random) : LocalDate = longProgression.random(random).let(LocalDate.Companion::fromEpochDays)
+
+/**
+ * Returns a random [LocalDate] within the bounds of the [LocalDateProgression] or null if the progression is empty.
+ *
+ * Takes the step into account; will not return any value within the range that would be skipped over by the progression.
+ *
+ * @sample kotlinx.datetime.test.samples.LocalDateRangeSamples.random
+ */
+public fun LocalDateProgression.randomOrNull(random: Random = Random) : LocalDate? = longProgression.randomOrNull(random)
+    ?.let(LocalDate.Companion::fromEpochDays)
```
These implementations can be made simpler (and more consistent with the other ranges) if they are limited to `LocalDateRange` instead of `LocalDateProgression`. Then, we know that `longProgression` is also a `LongRange`, so we can call `LongRange.random()`.

PeterAttardo replied:
While `longProgression` for a `LocalDateRange` is functionally identical to a `LongRange` it is not, as implemented, technically an instance of that class.

`Range`s are essentially `Progression`s with a step size of `1`, but there is nothing in the constructors or companion creation methods that forces a `Progression` with a step size of `1` to actually _be_ a `Range`. `LongProgression.fromClosedRange(start, end, 1)` will return an instance of `LongProgression`, not a `LongRange` downcast to a `LongProgression`.

We can check if the `step` is `1` and create a `LongRange` if so, but that logic doesn't seem any simpler than what we have now, for reduced functionality.

dkhalanskyjb replied:
The idea was to cast to a `LongRange`, not to check the `step`; also, with inheritance, we could also specialize `longProgression` to have the `LongRange` type, so the cast wouldn't even be needed.

That said, now that I think about it, since `LocalDateProgression` is a `Collection` now, adding `random` to it seems like a good idea, because otherwise, `random` will still work, but only because `Collection` itself has such an extension function. So, I agree we should keep `LocalDateProgression.random`.

dkhalanskyjb reviewed (changes requested):
I like how this is coming together!

dkhalanskyjb commented on core/common/src/LocalDateRange.kt:
```diff
@@ -0,0 +1,263 @@
+
+internal fun LongProgression.random(random: Random = Random) : Long = random.nextLong(0L..(last - first) / step) * step + first
+
+internal fun LongProgression.randomOrNull(random: Random = Random) : Long? = if (isEmpty()) null else random(random)
```
```suggestion
// incorrect in general; see `random` just above
private fun LongProgression.randomOrNull(random: Random = Random) : Long? = if (isEmpty()) null else random(random)
```

dkhalanskyjb commented on core/common/src/LocalDateRange.kt:
```diff
@@ -0,0 +1,263 @@
+
+internal fun LongProgression.randomOrNull(random: Random = Random) : Long? = if (isEmpty()) null else random(random)
+
+internal fun LongProgression.contains(element: Long) : Boolean = element in (if(step > 0) first..last else last..first) && (element - first) % step == 0L
```
```suggestion
// this implementation is incorrect in general (for example, `(Long.MIN_VALUE..Long.MAX_VALUE).step(5).contains(2)` returns `false`
// incorrectly https://www.wolframalpha.com/input?i=-2%5E63+%2B+1844674407370955162+*+5),
// but for the range of epoch days in LocalDate it's good enough
private fun LongProgression.contains(element: Long) : Boolean = element in (if(step > 0) first..last else last..first) && (element - first) % step == 0L
```

dkhalanskyjb commented on core/common/src/LocalDateRange.kt:
```diff
@@ -0,0 +1,263 @@
+ * Returns a [LocalDateProgression] with the same start and end, but a changed step value.
+ *
+ * **Pitfall**: the value parameter represents the magnitude of the step, not the direction, and therefore must be positive.
+ * Its sign will be matched to the sign of the existing step, in order to maintain the directionality of the progression.
```
> directionality

I don't see a suitable meaning in https://www.merriam-webster.com/dictionary/directionality
Maybe just "direction"?

PeterAttardo replied:
This feels like one of those cases where I have an intuitive feel of the nuances of English, but can't point to a hard and fast rule. A progression has directionality, as well as having _a_ direction. I think in this case either `directionality` or `direction` is valid, so I'll defer to whichever you prefer.

dkhalanskyjb replied:
My (non-native) intuitive feel of English tells me that ranges as a concept have directionality, but a specific range has a direction. Also, if both words can be used here, I'd stick to the simpler one so as not to distract the reader. API references should strive to be boring to read, I think.

dkhalanskyjb commented on core/common/src/LocalDateRange.kt:
```diff
@@ -0,0 +1,263 @@
+ * Returns a [LocalDateProgression] with the same start and end, but a changed step value.
+ *
+ * **Pitfall**: the value parameter represents the magnitude of the step, not the direction, and therefore must be positive.
+ * Its sign will be matched to the sign of the existing step, in order to maintain the directionality of the progression.
```
Ditto.

dkhalanskyjb commented on core/common/src/LocalDateRange.kt:
```diff
@@ -0,0 +1,263 @@
+
+/**
+ * Returns a reversed [LocalDateProgression], i.e. one that goes from [last] to [first].
+ * The sign of the step is switched, in order to reverse the directionality of the progression.
```
Ditto.

dkhalanskyjb commented on core/common/test/LocalDateRangeTest.kt:
```diff
@@ -0,0 +1,274 @@
+        assertEquals((Jan_01_2000..Jan_01_2000).last(), Jan_01_2000)
+        assertEquals((Jan_01_2000 downTo Jan_01_2000).last(), Jan_01_2000)
+        assertEquals((Jan_01_2000..Jan_05_2000).last(), Jan_05_2000)
+        assertEquals((Jan_05_2000 downTo Jan_01_2000).last(), Jan_01_2000)
```
Also worth testing that `last()` returns not the specified right boundary but the actual last emitted element, like `(1..10).step(5).last() == 6`.

dkhalanskyjb commented on core/common/src/LocalDateRange.kt:
```diff
@@ -0,0 +1,263 @@
+ *
+ * @sample kotlinx.datetime.test.samples.LocalDateRangeSamples.progressionWithStep
+ */
+public fun LocalDateProgression.step(value: Long, unit: DateTimeUnit.DayBased) : LocalDateProgression = LocalDateProgression(longProgression.step(safeMultiply(value, unit.days.toLong())))
```
I don't think `step` should throw an exception if `value * unit.days` overflows a `Long`. We can just clamp the result to `Long.(MAX|MIN)_VALUE` depending on the signs of the inputs, and `step` will behave correctly.

Or even if we decide to throw a `DateTimeArithmeticException`, we must at least declare it.

In any case, it's also a thing worth an additional test.

dkhalanskyjb commented on core/common/src/LocalDateRange.kt:
```diff
@@ -0,0 +1,263 @@
+    override fun toString(): String = "$first..$last"
+
+    public companion object {
+        public val EMPTY: LocalDateRange = LocalDateRange(LocalDate(1970, 1, 2), LocalDate(1970, 1, 1))
```
The documentation is missing.

dkhalanskyjb commented on core/common/src/LocalDateRange.kt:
```diff
@@ -0,0 +1,263 @@
+    override val endInclusive: LocalDate get() = last
+
+    /**
+     * returns the upper bound of the range, exclusive.
```
```suggestion
     * Returns the upper bound of the range, exclusive.
```

dkhalanskyjb reviewed (changes requested):
I noticed one important thing (`step()` throwing) and several small things. Once all of this is cleaned up, from my point of view, it's good to go! I added `@ilya-g`, as he may have useful insights, especially regarding the implementation of `Collection`.

PeterAttardo commented on core/common/src/LocalDateRange.kt:
```diff
@@ -0,0 +1,263 @@
+
+internal fun LongProgression.contains(element: Long) : Boolean = element in (if(step > 0) first..last else last..first) && (element - first) % step == 0L
+
+internal val LongProgression.size: Int
```
This looks to be an issue because the negative of `Long.MIN_VALUE` wraps back around to itself due to an overflow error. Do we want to solve this by implementing a `safeSubtract` method instead of using `safeAdd` on the negative of the subtrahend, or is it not worth the trouble, and we can get by with this explanatory comment?

dkhalanskyjb reviewed (changes requested):
The compilation fails for me, and when I fix it, the `getSize` test fails.

ilya-g commented on core/common/test/LocalDateRangeTest.kt:
```diff
@@ -0,0 +1,283 @@
+
+    @Test
+    fun reversed() {
+        assertContentEquals(
+            Jan_05_2000 downTo Jan_01_2000,
+            (Jan_01_2000..Jan_05_2000).reversed()
+        )
+        assertContentEquals(
+            Jan_01_2000..Jan_05_2000,
+            (Jan_05_2000 downTo Jan_01_2000).reversed()
+        )
+        assertContentEquals(
+            Jan_01_2000..Jan_01_2000,
+            (Jan_01_2000..Jan_01_2000).reversed()
+        )
```
I believe it should check more stricter - for equality, not for content equality

ilya-g commented on core/common/test/LocalDateRangeTest.kt:
```diff
@@ -0,0 +1,283 @@
+import kotlin.random.nextLong
+import kotlin.test.*
+
+class LocalDateRangeTest {
```
Could you also add tests for equals/hashCode contract of ranges/progressions?
E.g. that a progression with the step of 1 day is equal to a range. Or that a progression with the step of 7 days is equal to a progression with the step of a week.

dkhalanskyjb commented on core/common/src/LocalDateRange.kt:
```diff
@@ -0,0 +1,272 @@
+     * i.e. [value] is between [start] and [endInclusive].
+     */
+    @Suppress("ConvertTwoComparisonsToRangeCheck")
+    override fun contains(value: LocalDate): Boolean = first <= value && value <= last
```
There's a compiler warning that's actually helpful: `The corresponding parameter in the supertype 'LocalDateProgression' is named 'element'. This may cause problems when calling this function with named arguments.` I agree, let's rename this to `contains(element)`.

PeterAttardo replied:
I saw that earlier. Unfortunately `ClosedRange<T>` and `OpenEndRange<T>` both define the method `public operator fun contains(value: T): Boolean` while `Collection<E>` defines `public operator fun contains(element: E)`. `LocalDateRange` implements all three interfaces. They are all implemented with a single method, so whatever name one chooses for the parameter in that method is going to generate that warning for at least one of the parent interfaces.

ilya-g replied:
In the standard library, we chose to have the parameter named `value` both in range's and progression's `contains`

dkhalanskyjb reviewed (approved):
Other than the small naming issue, I think it's good to go!

ilya-g commented on core/common/test/LocalDateRangeTest.kt:
```diff
@@ -0,0 +1,312 @@
+            (Jan_01_2000 downTo Jan_01_2000).random()
+        )
+
+        assertFailsWith<IllegalArgumentException> {
```
Should be `NoSuchElementException`

PeterAttardo replied:
The underlying `Random.nextLong(range: LongRange)` call throws an `IllegalArgumentException` for an empty range. Do we want to wrap that call in a try-catch and rethrow as a `NoSuchElementException`?

ilya-g replied:
Yes, that's what `random()` function throws on empty collections and lists in stdlib.

ilya-g replied:
But it would be better to check for an empty range instead of wrapping a call to `nextLong()`

PeterAttardo commented on core/common/src/LocalDateRange.kt:
```diff
@@ -0,0 +1,272 @@
+    /**
+     * Returns true iff every element in [elements] is a member of the progression.
+     */
+    override fun containsAll(elements: Collection<LocalDate>): Boolean = elements.all(::contains)
```
Is this a deeper issue than at first appears? The signature of the methods implies no such unsafety, however after digging into it I can see that the methods in `AbstractCollection` are annotated with `@UnsafeVariance`, while the methods in `UByteArray` contain explicit type checks. I suppose it's a concession to Java compatibility where the `contains` methods take instances of `Object` instead of a typed parameter?

In any event, I'll copy the type check logic from `UByteArray` and add the test.

dkhalanskyjb commented:
Thank you a lot for the hard work! Merged manually in https://github.com/Kotlin/kotlinx-datetime/commit/d15ec21e1d53944168357bf70934fec46107730f to squash this into one commit.

lukellmann commented:
I noticed that `LocalDateProgression.random` [doesn't throw the documented exception](https://github.com/Kotlin/kotlinx-datetime/blob/06b389bb3a03ea97a41cf1662580bc2f00419c90/core/common/src/LocalDateRange.kt#L264-L269).

dkhalanskyjb commented:
@lukellmann, thanks, added the fix to the `year-month` branch (which, too, adds a `random` function): https://github.com/Kotlin/kotlinx-datetime/pull/457/commits/8c04d699e4fea5213b4fc4410f26aa8eb5c2c74a
